--- files.orig	2010-10-15 07:16:08.000000000 +0900
+++ files	2010-10-15 07:31:46.000000000 +0900
@@ -2226,8 +2226,8 @@
 net/mppcd.c			optional netgraph_mppc_compression
 net/netisr.c			standard
 net/pfil.c			optional ether | inet
-net/radix.c			standard
-net/radix_mpath.c		standard
+#net/ptree.c			standard
+net/ptree_mpath.c		standard
 net/raw_cb.c			standard
 net/raw_usrreq.c		standard
 net/route.c			standard
