--- if.c.org	2010-09-28 04:52:13.000000000 +0900
+++ if.c	2010-09-28 06:21:38.000000000 +0900
@@ -68,7 +68,7 @@
 #include <net/if_dl.h>
 #include <net/if_types.h>
 #include <net/if_var.h>
-#include <net/radix.h>
+#include <net/ptree.h>
 #include <net/route.h>
 #include <net/vnet.h>
 
@@ -132,7 +132,7 @@
 static int	if_transmit(struct ifnet *ifp, struct mbuf *m);
 static void	if_unroute(struct ifnet *, int flag, int fam);
 static void	link_rtrequest(int, struct rtentry *, struct rt_addrinfo *);
-static int	if_rtdel(struct radix_node *, void *);
+static int	if_rtdel(struct ptree_node *, void *);
 static int	ifhwioctl(u_long, struct ifnet *, caddr_t, struct thread *);
 static int	if_delmulti_locked(struct ifnet *, struct ifmultiaddr *, int);
 static void	do_link_state_change(void *, int);
@@ -807,7 +807,7 @@
 if_detach_internal(struct ifnet *ifp, int vmove)
 {
 	struct ifaddr *ifa;
-	struct radix_node_head	*rnh;
+	struct ptree	*rnh;
 	int i, j;
 	struct domain *dp;
  	struct ifnet *iter;
@@ -1326,7 +1326,7 @@
  *
  */
 static int
-if_rtdel(struct radix_node *rn, void *arg)
+if_rtdel(struct ptree_node *rn, void *arg)
 {
 	struct rtentry	*rt = (struct rtentry *)rn;
 	struct ifnet	*ifp = arg;
@@ -1695,7 +1695,7 @@
 				 * for an even better one.
 				 */
 				if (ifa_maybe == NULL ||
-				    rn_refines((caddr_t)ifa->ifa_netmask,
+				    ptree_refines((caddr_t)ifa->ifa_netmask,
 				    (caddr_t)ifa_maybe->ifa_netmask)) {
 					if (ifa_maybe != NULL)
 						ifa_free(ifa_maybe);
