--- route.h.orig	2005-01-11 01:52:56.000000000 +0900
+++ route.h	2005-02-05 04:29:23.000000000 +0900
@@ -124,20 +124,22 @@
  * gateway rather than the ultimate destination.
  */
 #ifndef RNF_NORMAL
-#include <net/radix.h>
+#include "opt_ptree.h"
+#include <net/ptree.h>
+#include <net/ptree_mpath.h>
 #ifdef RADIX_MPATH
 #include <net/radix_mpath.h>
 #endif
 #endif
 struct rtentry {
-	struct	radix_node rt_nodes[2];	/* tree glue, and other values */
+	struct	ptree_node *rt_nodes;	/* tree glue, and other values */
 	/*
 	 * XXX struct rtentry must begin with a struct radix_node (or two!)
 	 * because the code does some casts of a 'struct radix_node *'
 	 * to a 'struct rtentry *'
 	 */
-#define	rt_key(r)	(*((struct sockaddr **)(&(r)->rt_nodes->rn_key)))
-#define	rt_mask(r)	(*((struct sockaddr **)(&(r)->rt_nodes->rn_mask)))
+#define	rt_key(r)	(*((struct sockaddr **)(&(r)->rt_nodes->key)))
+#define	rt_mask(r)	(*((struct sockaddr **)(&(r)->rt_nodes->mask)))
 	struct	sockaddr *rt_gateway;	/* value */
 	int	rt_flags;		/* up/down?, host/net */
 	int	rt_refcnt;		/* # held references */
@@ -373,7 +375,7 @@
 	}							\
 } while (0)
 
-struct radix_node_head *rt_tables_get_rnh(int, int);
+struct ptree_node_head *rt_tables_get_rnh(int, int);
 
 struct ifmultiaddr;
 
