--- route.h.orig	2005-01-11 01:52:56.000000000 +0900
+++ route.h	2005-02-07 03:40:27.000000000 +0900
@@ -124,13 +124,15 @@
  * gateway rather than the ultimate destination.
  */
 #ifndef RNF_NORMAL
-#include <net/radix.h>
+#include "opt_ptree.h"
+#include <net/ptree.h>
+#include <net/ptree_mpath.h>
 #ifdef RADIX_MPATH
 #include <net/radix_mpath.h>
 #endif
 #endif
 struct rtentry {
-	struct	radix_node rt_nodes[2];	/* tree glue, and other values */
+	struct	ptree_node *rt_nodes;	/* tree glue, and other values */
 	/*
 	 * XXX struct rtentry must begin with a struct radix_node (or two!)
 	 * because the code does some casts of a 'struct radix_node *'
@@ -151,6 +153,10 @@
 #endif
 };
 
+struct rtentry_array {
+		struct rtentry **multi_entry;
+};
+
 /*
  * Following structure necessary for 4.3 compatibility;
  * We should eventually move it to a compat file.
@@ -373,7 +379,7 @@
 	}							\
 } while (0)
 
-struct radix_node_head *rt_tables_get_rnh(int, int);
+struct ptree_node_head *rt_tables_get_rnh(int, int);
 
 struct ifmultiaddr;
 
